include/deal.II/base/function_restriction.h

```diff
@@ -0,0 +1,160 @@
+
+
+
+  /**
+   * This class creates a 1-dimensional function from a dim + 1 dimension
+   * function by restriction dim of the coordinate values to a given point.
+   * Mathematically this corresponds to taking a function f = f(x, y, z), and a
+   * point (Y, Z), and defining a new function g = g(x) = f(x, Y, Z).
+   *
+   * The coordinates of the point will be expanded in the higher-dimensional
+   * functions coordinates starting from the open-direction (and wrapping
+   * around). In particular, if we restrict to a point (Z, X) and choose to keep
+   * the y-direction open, the restriction that is created is the function
+   * g(y) = f(X, y, Z).
+   * This is consistent with the convention in BoundingBox::cross_section.
+   */
+  template <int dim>
+  class PointRestriction : public Function<1>
+  {
+  public:
+    /**
+     * Constructor, takes the point that the incoming function should be
+     * restricted to and which (dim + 1)-dimensional coordiante direction should
+     * be kept "open".
+     *
+     * A pointer to the incoming function is stored internally, so the function
+     * must have a longer lifetime than the created restriction.
+     */
+    PointRestriction(const Function<dim + 1> &function,
+                     const unsigned int       open_direction,
+                     const Point<dim> &       point);
+
+    double
+    value(const Point<1> &point, const unsigned int component) const override;
+
+    Tensor<1, 1>
+    gradient(const Point<1> &   point,
+             const unsigned int component) const override;
+
+    SymmetricTensor<2, 1>
+    hessian(const Point<1> &point, const unsigned int component) const override;
+
+  private:
+    // The higher-dimensional function that has been restricted.
+    const SmartPointer<const Function<dim + 1>> function;
+
+    // The (dim + 1)-coordinate direction that is kept "open"
+    const unsigned int open_direction;
+
+    // The point that we have restricted the above function to.
+    const Point<dim> point;
+  };
+
+} // namespace Functions
+
+
+namespace internal
+{
+  /**
+   * Creates a (dim + 1)-dimensional point by copying over the coordinates of
+   * the incoming dim-dimensional point and setting the "missing"
+   * (dim + 1)-dimensional component to the incoming coordinate value.
+   *
+   * For example, given the input
+   * {(x, y), 2, z} this function creates the point (x, y, z).
+   *
+   * The coordinates of the dim-dimensional point are written to the coordinates
+   * of the (dim+1)-dimensional point in the order of the convention given by
+   * the function coordinate_to_one_dim_higher. Thus, the order of coordinates
+   * on the lower-dimensional point are not preserved:
+   * {(z, x), 1, y} creates the point (x, y, z).
+   */
+  template <int dim>
+  Point<dim + 1>
+  create_higher_dim_point(const Point<dim> & point,
+                          const unsigned int component_in_dim_plus_1,
+                          const double       coordinate_value);
```

Comment:
move this function into an anonymous namespace in the `.cc` file. It doesn't have to be declared in the header file and doesn't have to be exported from the .cc file either.

Reply:
I agree that it doesn't make sense to have it header here, but I reuse this function in the implementation of the algorithm that I mention above. So I would like to have it exposed somewhere.